File: lib/features/player/models/player_battlelog.dart
import 'package:clashkingapp/core/constants/image_assets.dart';
import 'package:clashkingapp/core/services/game_data_service.dart';

enum PlayerBattlelogMode { ranked, farming }

enum PlayerBattlelogSource { official, history }

class PlayerBattlelogEntry {
  const PlayerBattlelogEntry({
    required this.id,
    required this.mode,
    required this.source,
    required this.attack,
    required this.opponentTag,
    required this.opponentName,
    required this.opponentTownHall,
    required this.stars,
    required this.destructionPercentage,
    required this.gold,
    required this.elixir,
    required this.darkElixir,
    required this.timestamp,
    required this.duration,
    required this.armyShareCode,
    required this.armyCounts,
  });

  final String id;
  final PlayerBattlelogMode mode;
  final PlayerBattlelogSource source;
  final bool attack;
  final String opponentTag;
  final String opponentName;
  final int opponentTownHall;
  final int stars;
  final int destructionPercentage;
  final int gold;
  final int elixir;
  final int darkElixir;
  final DateTime? timestamp;
  final int duration;
  final String armyShareCode;
  final Map<String, int> armyCounts;

  int get totalLoot => gold + elixir + darkElixir;

  factory PlayerBattlelogEntry.fromOfficial(Map<String, dynamic> json) {
    final resources = <String, int>{};
    for (final raw in _list(json['lootedResources'])) {
      final resource = _map(raw);
      final name = resource['name']?.toString().toLowerCase() ?? '';
      resources[name] = _int(resource['amount']);
    }
    final armyShareCode = json['armyShareCode']?.toString() ?? '';
    final timestamp = _parseBattleTime(json['battleTimestamp']);
    final opponentTag = json['opponentPlayerTag']?.toString() ?? '';
    return PlayerBattlelogEntry(
      id: '',
      mode: _battlelogMode(json['battleType']),
      source: PlayerBattlelogSource.official,
      attack: json['attack'] == true,
      opponentTag: opponentTag,
      opponentName: json['opponentName']?.toString() ?? '',
      opponentTownHall: _int(json['opponentTownHallLevel']),
      stars: _int(json['stars']),
      destructionPercentage: _int(json['destructionPercentage']),
      gold: resources['gold'] ?? 0,
      elixir: resources['elixir'] ?? 0,
      darkElixir: resources['darkelixir'] ?? resources['dark_elixir'] ?? 0,
      timestamp: timestamp,
      duration: _int(json['battleTime']),
      armyShareCode: armyShareCode,
      armyCounts: _parseArmyCounts(armyShareCode),
    );
  }

  factory PlayerBattlelogEntry.fromHistory(Map<String, dynamic> json) {
    final armyShareCode = json['army_share_code']?.toString() ?? '';
    final storedArmyCounts = _map(
      json['army_counts'],
    ).map((key, value) => MapEntry(key, _int(value)));
    return PlayerBattlelogEntry(
      id: json['battle_id']?.toString() ?? '',
      mode: _battlelogMode(json['battle_type']),
      source: PlayerBattlelogSource.history,
      attack: json['attack'] == true,
      opponentTag: json['opponent_tag']?.toString() ?? '',
      opponentName: json['opponent_name']?.toString() ?? '',
      opponentTownHall: _int(json['opponent_townhall']),
      stars: _int(json['stars']),
      destructionPercentage: _int(json['destruction_percentage']),
      gold: _int(json['gold']),
      elixir: _int(json['elixir']),
      darkElixir: _int(json['dark_elixir']),
      timestamp: _parseBattleTime(json['timestamp']),
      duration: _int(json['duration']),
      armyShareCode: armyShareCode,
      armyCounts: storedArmyCounts.isEmpty
          ? _parseArmyCounts(armyShareCode)
          : storedArmyCounts,
    );
  }

  String get mergeKey {
    final seconds = timestamp?.toUtc().millisecondsSinceEpoch ?? 0;
    return '${attack ? 1 : 0}|${opponentTag.toUpperCase()}|$seconds';
  }
}

class PlayerBattlelogData {
  const PlayerBattlelogData({
    required this.items,
    required this.officialAvailable,
    required this.historyAvailable,
  });

  final List<PlayerBattlelogEntry> items;
  final bool officialAvailable;
  final bool historyAvailable;

  List<PlayerBattlelogEntry> forMode(PlayerBattlelogMode mode) =>
      items.where((item) => item.mode == mode).toList(growable: false);

  static PlayerBattlelogData merge({
    required Iterable<PlayerBattlelogEntry> official,
    required Iterable<PlayerBattlelogEntry> history,
    required bool officialAvailable,
    required bool historyAvailable,
  }) {
    final merged = <String, PlayerBattlelogEntry>{};
    for (final item in official) {
      merged[item.mergeKey] = item;
    }
    // Historical rows carry parsed armies, so they intentionally win when the
    // official endpoint returns the same battle.
    for (final item in history) {
      merged[item.mergeKey] = item;
    }
    final items = merged.values.toList()
      ..sort((a, b) {
        final left = a.timestamp ?? DateTime.fromMillisecondsSinceEpoch(0);
        final right = b.timestamp ?? DateTime.fromMillisecondsSinceEpoch(0);
        return right.compareTo(left);
      });
    return PlayerBattlelogData(
      items: items,
      officialAvailable: officialAvailable,
      historyAvailable: historyAvailable,
    );
  }

  List<PlayerPopularArmyItem> popularTroops(
    PlayerBattlelogMode mode, {
    int limit = 3,
  }) {
    final uses = <String, int>{};
    for (final battle in items.where(
      (item) => item.mode == mode && item.attack,
    )) {
      for (final code in battle.armyCounts.keys.where(
        (code) => code.startsWith('u_') || code.startsWith('i_'),
      )) {
        uses.update(code, (count) => count + 1, ifAbsent: () => 1);
      }
    }
    final popular =
        uses.entries
            .map(
              (entry) => PlayerPopularArmyItem(
                item: PlayerBattlelogArmyCatalog.resolve(entry.key),
                uses: entry.value,
              ),
            )
            .toList()
          ..sort((a, b) {
            final byUses = b.uses.compareTo(a.uses);
            return byUses != 0 ? byUses : a.item.name.compareTo(b.item.name);
          });
    return popular.take(limit).toList(growable: false);
  }
}

class PlayerPopularArmyItem {
  const PlayerPopularArmyItem({required this.item, required this.uses});

  final PlayerBattlelogArmyItem item;
  final int uses;
}

class PlayerBattlelogArmyItem {
  const PlayerBattlelogArmyItem({
    required this.code,
    required this.name,
    required this.imageUrl,
  });

  final String code;
  final String name;
  final String imageUrl;
}

class PlayerBattlelogArmyCatalog {
  PlayerBattlelogArmyCatalog._();

  static PlayerBattlelogArmyItem resolve(String code) {
    final split = code.split('_');
    final prefix = split.firstOrNull ?? '';
    final itemID = split.length == 2 ? int.tryParse(split[1]) : null;
    final item = itemID == null ? null : _findItem(prefix, itemID);
    final name = GameDataService.localizedNameForItem(item).trim();
    final resolvedName = name.isEmpty ? code : name;
    final imageUrl = switch (prefix) {
      's' || 'd' => ImageAssets.getSpellImage(resolvedName),
      'h' => ImageAssets.getHeroImage(resolvedName),
      'p' => ImageAssets.getPetImage(resolvedName),
      'e' => ImageAssets.getGearImage(resolvedName),
      _ => ImageAssets.getTroopImage(resolvedName),
    };
    return PlayerBattlelogArmyItem(
      code: code,
      name: resolvedName,
      imageUrl: imageUrl,
    );
  }

  static Map<String, dynamic>? _findItem(String prefix, int itemID) {
    final sectionName = switch (prefix) {
      's' || 'd' => 'spells',
      'h' => 'heroes',
      'p' => 'pets',
      'e' => 'equipment',
      _ => 'troops',
    };
    final rawSection = GameDataService.bundleData[sectionName];
    for (final item in _items(rawSection)) {
      final rawID = _int(item['_id']);
      if (rawID == itemID || rawID % 1000000 == itemID) return item;
    }

    final normalized = switch (sectionName) {
      'spells' => GameDataService.spellsData['spells'],
      'heroes' => GameDataService.heroesData['heroes'],
      'pets' => GameDataService.petsData['pets'],
      'equipment' => GameDataService.gearsData['gears'],
      _ => GameDataService.troopsData['troops'],
    };
    for (final item in _items(normalized)) {
      final rawID = _int(item['_id']);
      if (rawID == itemID || rawID % 1000000 == itemID) return item;
    }
    return null;
  }

  static Iterable<Map<String, dynamic>> _items(Object? section) sync* {
    if (section is List) {
      for (final item in section) {
        if (item is Map) yield Map<String, dynamic>.from(item);
      }
    } else if (section is Map) {
      for (final item in section.values) {
        if (item is Map) yield Map<String, dynamic>.from(item);
      }
    }
  }
}

PlayerBattlelogMode _battlelogMode(Object? value) {
  final normalized = value?.toString().trim().toLowerCase().replaceAll(
    RegExp(r'[^a-z]'),
    '',
  );
  return switch (normalized) {
    'ranked' || 'legend' => PlayerBattlelogMode.ranked,
    'homevillage' || 'farming' => PlayerBattlelogMode.farming,
    _ => PlayerBattlelogMode.farming,
  };
}

Map<String, int> _parseArmyCounts(String shareCode) {
  final payload = Uri.tryParse(shareCode)?.queryParameters['army'] ?? shareCode;
  final counts = <String, int>{};
  for (final section in RegExp(r'([hidsu])([^hidsu]*)').allMatches(payload)) {
    final prefix = section.group(1)!;
    if (prefix == 'h') continue;
    for (final rawItem in section.group(2)!.split('-')) {
      final match = RegExp(r'^(\d+)x(\d+)$').firstMatch(rawItem);
      if (match == null) continue;
      final quantity = int.tryParse(match.group(1)!);
      final itemID = int.tryParse(match.group(2)!);
      if (quantity == null || quantity <= 0 || itemID == null) continue;
      counts.update(
        '${prefix}_$itemID',
        (current) => current + quantity,
        ifAbsent: () => quantity,
      );
    }
  }
  return counts;
}

DateTime? _parseBattleTime(Object? value) {
  final raw = value?.toString().trim() ?? '';
  if (raw.isEmpty) return null;
  final parsed = DateTime.tryParse(raw);
  if (parsed != null) return parsed.toUtc();
  final match = RegExp(
    r'^(\d{4})(\d{2})(\d{2})T(\d{2})(\d{2})(\d{2})(?:\.(\d+))?Z$',
  ).firstMatch(raw);
  if (match == null) return null;
  final milliseconds = (match.group(7) ?? '').padRight(3, '0');
  return DateTime.utc(
    int.parse(match.group(1)!),
    int.parse(match.group(2)!),
    int.parse(match.group(3)!),
    int.parse(match.group(4)!),
    int.parse(match.group(5)!),
    int.parse(match.group(6)!),
    int.tryParse(milliseconds.substring(0, 3)) ?? 0,
  );
}

Map<String, dynamic> _map(Object? value) =>
    value is Map ? Map<String, dynamic>.from(value) : const <String, dynamic>{};

List<dynamic> _list(Object? value) => value is List ? value : const [];

int _int(Object? value) =>
    value is num ? value.toInt() : int.tryParse(value?.toString() ?? '') ?? 0;

File: expo/src/features/player/models/player-battlelog.ts
import { ImageAssets } from '@/core/assets/image-assets';
import { localizedNameForItem } from '@/core/game-data/game-data-localization';
import { gameDataState } from '@/core/game-data/game-data-state';
import { apiDate, int, isRecord, record, records, string, type JsonRecord } from './parsing';

export type PlayerBattlelogMode = 'ranked' | 'farming';
export type PlayerBattlelogSource = 'official' | 'history';
export class PlayerBattlelogEntry {
  constructor(
    readonly id: string,
    readonly mode: PlayerBattlelogMode,
    readonly source: PlayerBattlelogSource,
    readonly attack: boolean,
    readonly opponentTag: string,
    readonly opponentName: string,
    readonly opponentTownHall: number,
    readonly stars: number,
    readonly destructionPercentage: number,
    readonly gold: number,
    readonly elixir: number,
    readonly darkElixir: number,
    readonly timestamp: Date | null,
    readonly duration: number,
    readonly armyShareCode: string,
    readonly armyCounts: Readonly<Record<string, number>>,
  ) {}
  get totalLoot() {
    return this.gold + this.elixir + this.darkElixir;
  }
  get mergeKey() {
    return `${this.attack ? 1 : 0}|${this.opponentTag.toUpperCase()}|${this.timestamp?.getTime() ?? 0}`;
  }
  static fromOfficial(json: JsonRecord) {
    const resources: Record<string, number> = {};
    for (const item of records(json.lootedResources)) {
      resources[string(item.name).toLowerCase().replace(/[\s_]/g, '')] = int(item.amount);
    }
    const share = string(json.armyShareCode);
    return new PlayerBattlelogEntry(
      '',
      battlelogMode(json.battleType),
      'official',
      json.attack === true,
      string(json.opponentPlayerTag),
      string(json.opponentName),
      int(json.opponentTownHallLevel),
      int(json.stars),
      int(json.destructionPercentage),
      resources.gold ?? 0,
      resources.elixir ?? 0,
      resources.darkelixir ?? 0,
      apiDate(json.battleTimestamp),
      int(json.battleTime),
      share,
      parseArmyCounts(share),
    );
  }
  static fromHistory(json: JsonRecord) {
    const share = string(json.army_share_code);
    const stored = Object.fromEntries(
      Object.entries(record(json.army_counts)).map(([key, value]) => [key, int(value)]),
    );
    return new PlayerBattlelogEntry(
      string(json.battle_id),
      battlelogMode(json.battle_type),
      'history',
      json.attack === true,
      string(json.opponent_tag),
      string(json.opponent_name),
      int(json.opponent_townhall),
      int(json.stars),
      int(json.destruction_percentage),
      int(json.gold),
      int(json.elixir),
      int(json.dark_elixir),
      apiDate(json.timestamp),
      int(json.duration),
      share,
      Object.keys(stored).length ? stored : parseArmyCounts(share),
    );
  }
}
export class PlayerBattlelogData {
  constructor(
    readonly items: readonly PlayerBattlelogEntry[],
    readonly officialAvailable: boolean,
    readonly historyAvailable: boolean,
  ) {}
  forMode(mode: PlayerBattlelogMode) {
    return this.items.filter((item) => item.mode === mode);
  }
  static merge(input: {
    official: Iterable<PlayerBattlelogEntry>;
    history: Iterable<PlayerBattlelogEntry>;
    officialAvailable: boolean;
    historyAvailable: boolean;
  }) {
    const merged = new Map<string, PlayerBattlelogEntry>();
    for (const item of input.official) merged.set(item.mergeKey, item);
    for (const item of input.history) merged.set(item.mergeKey, item);
    return new PlayerBattlelogData(
      [...merged.values()].sort(
        (a, b) => (b.timestamp?.getTime() ?? 0) - (a.timestamp?.getTime() ?? 0),
      ),
      input.officialAvailable,
      input.historyAvailable,
    );
  }
  popularTroops(mode: PlayerBattlelogMode, options: { limit?: number; attack?: boolean } = {}) {
    const uses = new Map<string, number>();
    for (const battle of this.items.filter(
      (item) => item.mode === mode && item.attack === (options.attack ?? true),
    ))
      for (const code of Object.keys(battle.armyCounts).filter(
        (code) => code.startsWith('u_') || code.startsWith('i_'),
      ))
        uses.set(code, (uses.get(code) ?? 0) + 1);
    return [...uses]
      .map(
        ([code, count]) =>
          new PlayerPopularArmyItem(PlayerBattlelogArmyCatalog.resolve(code), count),
      )
      .sort((a, b) => b.uses - a.uses || a.item.name.localeCompare(b.item.name))
      .slice(0, options.limit ?? 3);
  }
}
export class PlayerPopularArmyItem {
  constructor(
    readonly item: PlayerBattlelogArmyItem,
    readonly uses: number,
  ) {}
}
export class PlayerBattlelogArmyItem {
  constructor(
    readonly code: string,
    readonly name: string,
    readonly imageUrl: string,
  ) {}
}
export class PlayerBattlelogArmyCatalog {
  static resolve(code: string) {
    const [prefix = '', rawId] = code.split('_');
    const id = rawId === undefined ? null : Number.parseInt(rawId, 10);
    const item = id === null || Number.isNaN(id) ? null : findItem(prefix, id);
    const translated = localizedNameForItem(item).trim();
    const name = translated || code;
    const image =
      prefix === 's' || prefix === 'd'
        ? ImageAssets.getSpellImage(name)
        : prefix === 'h'
          ? ImageAssets.getHeroImage(name)
          : prefix === 'p'
            ? ImageAssets.getPetImage(name)
            : prefix === 'e'
              ? ImageAssets.getGearImage(name)
              : ImageAssets.getTroopImage(name);
    return new PlayerBattlelogArmyItem(code, name, image);
  }
}
function findItem(prefix: string, id: number): JsonRecord | null {
  const section =
    prefix === 's' || prefix === 'd'
      ? 'spells'
      : prefix === 'h'
        ? 'heroes'
        : prefix === 'p'
          ? 'pets'
          : prefix === 'e'
            ? 'equipment'
            : 'troops';
  const normalized =
    section === 'spells'
      ? gameDataState.spellsData.spells
      : section === 'heroes'
        ? gameDataState.heroesData.heroes
        : section === 'pets'
          ? gameDataState.petsData.pets
          : section === 'equipment'
            ? gameDataState.gearsData.gears
            : gameDataState.troopsData.troops;
  for (const source of [gameDataState.bundleData[section], normalized])
    for (const item of sectionItems(source)) {
      const raw = int(item._id);
      if (raw === id || raw % 1_000_000 === id) return item;
    }
  return null;
}
function sectionItems(value: unknown): JsonRecord[] {
  if (Array.isArray(value)) return value.filter(isRecord);
  return isRecord(value) ? Object.values(value).filter(isRecord) : [];
}
function battlelogMode(value: unknown): PlayerBattlelogMode {
  const normalized = string(value)
    .trim()
    .toLowerCase()
    .replace(/[^a-z]/g, '');
  return normalized === 'ranked' || normalized === 'legend' ? 'ranked' : 'farming';
}
export function parseArmyCounts(shareCode: string): Record<string, number> {
  let payload = shareCode;
  try {
    payload = new URL(shareCode).searchParams.get('army') ?? shareCode;
  } catch {
    const query = shareCode.match(/[?&]army=([^&]+)/)?.[1];
    if (query) payload = decodeURIComponent(query);
  }
  const counts: Record<string, number> = {};
  for (const section of payload.matchAll(/([hidsu])([^hidsu]*)/g)) {
    const prefix = section[1]!;
    if (prefix === 'h') continue;
    for (const rawItem of section[2]!.split('-')) {
      const match = /^(\d+)x(\d+)$/.exec(rawItem);
      if (!match) continue;
      const quantity = Number(match[1]),
        id = Number(match[2]);
      if (quantity <= 0) continue;
      const key = `${prefix}_${id}`;
      counts[key] = (counts[key] ?? 0) + quantity;
    }
  }
  return counts;
}
